specification/sbi_atee.adoc

```diff
@@ -10,8 +39,1238 @@ shown in the figure above, the TEE Runtime (operating in VS-mode) supports
+| SBI_ERR_FAILED          | The operation failed for unknown reasons.
+|===
+
+=== Function: TEE Host Finalize TVM (FID #6)
+[source, C]
+------
+struct sbiret sbi_tee_host_finalize_tvm(unsigned long tvm_guest_id);
+------
+Transitions the TVM specified by `tvm_guest_id` from the "TVM_INITIALIZING" state to a "TVM_RUNNABLE"
+state. The host must finalize TVM shared-memory execution parameters like the entry point (`ENTRY_PC`)
+and boot argument (`ENTRY_ARG`) on the boot VCPU configured by `sbi_tee_host_create_tvm()` before making
+this call. 
+
+The possible error codes returned in `sbiret.error` are shown below.
+
+[#table_sbi_tee_host_finalize_tvm_errors]
+.TEE Host Finalize TVM Errors
+[cols="2,3", width=90%, align="center", options="header"]
+|===
+| Error code            | Description
+| SBI_SUCCESS           | The operation completed successfully.
+| SBI_ERR_INVALID_PARAM | `tvm_guest_id` was invalid, or the
+                          TVM wasn't in the "initializing" state.
+| SBI_ERR_FAILED        | The operation failed for unknown reasons.
+|===
+ 
+=== Function: TEE Host Destroy TVM (FID #7)
+[source, C]
+-------
+struct sbiret sbi_tee_host_destroy_tvm(unsigned long tvm_guest_id);
+-------
+
+Destroys a confidential TVM previously created using *`sbi_tee_host_create_tvm()`*.  
+
+Confidential TVM memory is automatically released following successful destruction, and it
+can be assigned to other TVMs. Repurposing confidential memory for use by non-confidential
+TVMs requires an explicit call to *`sbi_tee_reclaim_pages()`* (described below).
+
+The possible error codes returned in `sbiret.error` are shown below.
+
+[#table_sbi_tee_host_destroy_tvm_errors]
+.TEE Host Destroy TVM Errors
+[cols="2,3", width=90%, align="center", options="header"]
+|===
+| Error code            | Description
+| SBI_SUCCESS           | The operation completed successfully.
+| SBI_ERR_INVALID_PARAM | `tvm_guest_id` was invalid.
+| SBI_ERR_FAILED        | The operation failed for unknown reasons.
+|===
+
+=== Function: TEE Host Add TVM Memory Region (FID #8)
+[source, C]
+-----
+struct sbiret sbi_tee_host_add_tvm_memory_region(unsigned long tvm_guest_id,
+                                                 unsigned long tvm_gpa_addr,
+                                                 unsigned long tvm_memory_region_type,
+                                                 unsigned long region_len);
+-----
+Marks the range of TVM physical address space starting at `tvm_gpa_addr` as reserved
+for the mapping of memory. The type of memory is specified by `tvm_memory_region_type`
+and the length is specified by by `region_len`. `tvm_memory_region_type` must be a
+legal value for the `tvm_memory_region_type` enum type described below.
+
+Both `tvm_gpa_addr` and `region_len` must be 4kB-aligned, and the region must not
+overlap with a previously defined region. This call must not be made after calling
+`sbi_tee_host_finalize_tvm()`.
```

Comment:
do we need this - once we add support for dynamic memory conversion?

Reply:
Looks like I somehow missed the conversation around dynamic memory (there probably wasn't one, except for Andrew's note about the document in the status last week). Let me take a look.

Reply:
Seems like the document calls for the API to be retained? cc: @abrestic-rivos 

```
_TEE-H changes:
teeh_add_mem_region(): Same as existing behavior except that the guest may later convert parts of the region to shared, thus the host should call this for the full GPA range of the host that may be populated. The ranges added here should be in the “memory” node of the TVM’s device tree._
```

Reply:
@rsahita Do you mean TSM should automatically add the entire memory nodes specified in the guest DT ?

Reply:
@atulkharerivos : It probably doesn't matter much as this API may change. But the current function signature doesn't match the implementation. 

Implementation:
TvmAddMemoryRegion {
        /// a0 = guest id
        guest_id: u64,
        /// a1 = type of memory region
        region_type: TeeMemoryRegion,
        /// a2 = start of the region
        guest_addr: u64,
        /// a3 = length of the region
        len: u64,
    },

Reply:
This will eventually become:

```
sbi_teeh_add_memory_region(tvm_id, tvm_addr, len)
```

and will only add confidential memory regions. Shared/MMIO are initiated by the guest.

Reply:
We should finish the discussion about sharing API with nested VMs before comitting to _only_ confidential memory support here. Could handle that with a separate API for adding shared memory, but that feels like a step backwards.

Reply:
Is there a good reason to even bother with memory regions for non-TEE nested virtualization though? The guest hypervisor should be allowed to insert pages or handle page faults in whatever way it wants for non-TVMs; requiring it to declare valid GPA ranges up front imposes a TEE-specific requirement that wouldn't otherwise be present in the non-nested case.

Reply:
Good point. As long as we have a path that allows mapping outside of a pre-registered region that should work fine.